submariner/pluggable-network-plugin-and-ovn.md

```diff
@@ -74,6 +70,16 @@ These network-plugin specific drivers will interface to the underlying network-p
 and prepare it's configuration as necessary, as well as any routing or linux network
 stack changes necessary on the worker/gateway nodes of the cluster.
 
+### Synchronizer implementation
+
+Some parts of the drivers could need a single worker handling specific synchronization
+work with the network plugin.
+
+The drivers declare in a flag if they need this type of instance.
+
+The synchronizer instances receive the same exact events, but instead of configuring
```

Comment:
Two questions @mangelajo 

1. If we are going to spawn a new deployment (with replica count as 1), will we use the same image for that?

2. When you say that synchronization instance will receive the same exact events. Will it receive **all the events** listed above? Won't it be an overkill for the benefit of having it inside the driver itself (i.e., allow the driver implementation to use its own leader election and remove the leader events)

Reply:
> Two questions @mangelajo
> 
> 1. If we are going to spawn a new deployment (with replica count as 1), will we use the same image for that?

We could use the same image/binary, just on a different mode, so we don't need to have two different type of drivers/interfaces. As they would receive the same settings etc...

but just an idea, trying to make things simpler. We can always split to more images/binaries if we need.

> 2. When you say that synchronization instance will receive the same exact events. Will it receive **all the events** listed above? Won't it be an overkill for the benefit of having it inside the driver itself (i.e., allow the driver implementation to use its own leader election and remove the leader events)


I think some of the events would become irrelevant for OVN on that synchronizer instance, we need to figure those details out. May be there are events that we can filter if the driver don't need them (in a later optimization phase?) but I think we may want to keep things as simple as possible if we want to have something this sprint.

May be we should have a meeting with @mkolesnik about these topics :-)

Reply:
As you know, creating a new deployment has implications on the installer tools like operator and helm. Also, if we use the same image, then within the driver code, we have to include checks if its doing regular event processing or if its purpose is to program the CNI. If we use a different image, then we will have associated issues like versioning, mis-matches in driver vs new image etc etc

I'm not against this proposal, but IMHO I get a feeling that splitting the route-agent driver into two pods will create more maintenance headaches. 

If we have the code as part of driver itself, it can use a leader election (if required) or the route-agent running on gatewayNode can act as leader. BTW, figuring out the active leader is very easy if we use "kubectl describe cm <configMapName>" - the output includes information about the active leader holding the lock.

Lets discuss with @mkolesnik for his inputs.